docs/csharp/write-safe-efficient-code.md

```diff
@@ -64,6 +64,8 @@ readonly public struct ReadonlyPoint3D
 
 Follow this recommendation whenever your design intent is to create an immutable value type. Any performance improvements are an added benefit. The `readonly struct` clearly expresses your design intent.
 
+The [unamanaged types](language-reference/keywords/sizeof.md) aren't explicitly marked as a `readonly struct` types, but the compiler treats them as such. These types are considered `readonly` from the perspective of the language. We know them to be immutable which is even a stronger guarantee than `readonly` can provide.
```

Comment:
The `readonliness` of a struct call is decided here: https://github.com/dotnet/roslyn/blob/60fd993cdab5a4a92490755422984cdfea752be8/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L1687

We special-case 3 methods on `Nullable<T>`, but otherwise we just honor `readonly` annotation on the type that implements the method. Treating primitive types as unconditionally readonly was considered, but I do not think it made it into spec or implementation.

I am not sure it is a very important distinction anyways - If types are truly immutable, cloning is not semantically observable and perf penalty, if any, is very small since these types are very small.
Basically only people that examine actual `IL` might see the difference, which makes this more a curiosity point, not something that would guide you to write your code one way instead of another.